scripts/irods/controller.py

```diff
@@ -188,27 +181,42 @@ def irods_grid_shutdown(self, timeout=20, **kwargs):
         else:
             raise IrodsError(
                 'The iRODS server did not stop within {0} seconds of '
-                'receiving the "shutdown" command.'.format(timeout))
+                'receiving SIGTERM.'.format(timeout))
 
     def stop(self, timeout=20):
         l = logging.getLogger(__name__)
         self.config.clear_cache()
         l.debug('Calling stop on IrodsController')
         l.info('Stopping iRODS server...')
         try:
-            if self.get_binary_to_pids_dict([self.config.server_executable]):
+            server_pid = lib.get_server_pid()
+            server_proc = None
+            if server_pid >= 0:
+                # make sure it's not a reused pid
+                server_proc = psutil.Process(server_pid)
+                if psutil.version_info >= (2, 0):
+                    server_pid_exec = server_proc.exe()
+                else:
+                    server_pid_exec = server_proc.exe
+                if server_pid_exec != self.config.server_executable:
+                    server_pid = -1
+                    server_proc = None
+
+            if server_pid >= 0:
```

Comment:
is this line redundant?    consider adding a comment or four in here for the next spelunker.

Reply:
The previous `if` block is a sanity check. Since `lib.server_get_pid()` does not have access to `self.config`, it can't check the pid from the pidfile against `self.config.server_executable`, only that a process with that pid exists. We check it here to make sure that the pid wasn't recycled for some other process. I've already left a comment about this on line 195.

The `else` block here handles the cases in which `server_pid` was negative going into the previous `if` block *and* in which `server_pid` was *set* negative in the previous `if` block. Since we need to check `server_pid` again anyway, it's neater to structure the two if blocks like this. One that double-checks that the return value from `lib.get_server_pid()` is valid and one that actually does the thing.

Reply:
indeed - on all counts.